tests/core/test_variable.py

```diff
@@ -0,0 +1,558 @@
+        self.assertAlmostEqual(val, 3.0)
+        self.assertEqual(state, v.State.FREE)
+
+        #construction with low and high bounds
+        v = relentless.DesignVariable(value=0.0, low=0.0, high=2.0)
+        #test below low
+        val,state = v.clamp(-1.0)
+        self.assertAlmostEqual(val, 0.0)
+        self.assertEqual(state, v.State.LOW)
+        #test between bounds
+        val,state = v.clamp(1.0)
+        self.assertAlmostEqual(val, 1.0)
+        self.assertEqual(state, v.State.FREE)
+        #test above high
+        val,state = v.clamp(2.5)
+        self.assertAlmostEqual(val, 2.0)
+        self.assertEqual(state, v.State.HIGH)
+
+        #construction with no bounds
+        v = relentless.DesignVariable(value=0.0)
+        #test free variable
+        val,state = v.clamp(1.0)
+        self.assertAlmostEqual(val, 1.0)
+        self.assertEqual(state, v.State.FREE)
+
+    def test_value(self):
+        """Test methods for setting values and checking bounds."""
+        #test construction with value between bounds
+        v = relentless.DesignVariable(value=0.0, low=-1.0, high=1.0)
+        self.assertAlmostEqual(v.value, 0.0)
+        self.assertEqual(v.state, v.State.FREE)
+
+        #test below low
+        v.value = -1.5
+        self.assertAlmostEqual(v.value, -1.0)
+        self.assertEqual(v.state, v.State.LOW)
+
+        #test above high
+        v.value = 3
+        self.assertAlmostEqual(v.value, 1.0)
+        self.assertEqual(v.state, v.State.HIGH)
+
+        #test invalid value
+        with self.assertRaises(ValueError):
+            v.value = '0'
+
+    def test_bounds(self):
+        """Test methods for setting bounds and checking value clamping."""
+        #test construction with value between bounds
+        v = relentless.DesignVariable(value=0.0, low=-1.0, high=1.0)
+        self.assertAlmostEqual(v.value, 0.0)
+        self.assertEqual(v.state, v.State.FREE)
+
+        #test changing low bound to above value
+        v.low = 0.2
+        self.assertAlmostEqual(v.value, 0.2)
+        self.assertAlmostEqual(v.low, 0.2)
+        self.assertAlmostEqual(v.high, 1.0)
+        self.assertEqual(v.state, v.State.LOW)
+
+        #test changing high bound to below value
+        v.low = -1.0
+        v.high = -0.2
+        self.assertAlmostEqual(v.value, -0.2)
+        self.assertAlmostEqual(v.low, -1.0)
+        self.assertAlmostEqual(v.high, -0.2)
+        self.assertEqual(v.state, v.State.HIGH)
+
+        #test changing bounds incorrectly
+        with self.assertRaises(ValueError):
+            v.low = 0.2
+        with self.assertRaises(ValueError):
+            v.high = -1.6
+
+class DepVar(relentless.DependentVariable):
+    """Mock dependent variable to test relentless.DependentVariable"""
+    def __init__(self, *a, **b):
+        super().__init__(*a, **b)
+
+    @property
+    def value(self):
+        return np.sum([v.value for p,v in self.depends])
+
+    def _derivative(self, param):
```

Comment:
Can you add a comment here that these derivatives aren't the actual derivs and that they are just used for testing? It might confused someone later.

Alternatively, do any of the tests ever trigger the first condition? You could also just modify these so that value and _derivative are compatible with each other.